File: lab2_task2/main.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

class Task2 {
    static int CountValidEmails(List<string> emails) {
        var validEmails = new HashSet<string>();
        var pattern = new Regex(@"^([a-zA-Z0-9\.]{6,30})(\*[a-zA-Z0-9\.]*)?@([a-zA-Z0-9\-\.]+)$");
        
        foreach (var email in emails) {
            var match = pattern.Match(email);
            if (!match.Success) {
                Console.WriteLine($"Неверная почта: {email}");
                continue;
            }
            
            var username = match.Groups[1].Value;
            var domain = match.Groups[3].Value;
            
            if (username.StartsWith(".") || username.EndsWith(".") || username.Contains("..")) {
                Console.WriteLine($"Лишние точки в имени: {email}");
                continue;
            }
            
            if (domain.StartsWith(".") || domain.EndsWith(".") || domain.Contains("..")) {
                Console.WriteLine($"Лишние точки в домене: {email}");
                continue;
            }
            
            var cleanUsername = username.Replace(".", "");
            if (cleanUsername.Contains("*")) {
                cleanUsername = cleanUsername.Split('*')[0];
                if (cleanUsername.Length < 6) {
                    Console.WriteLine($"Имя пользователя слишком короткое: {email}");
                    continue;
                }
            }
            
            validEmails.Add($"{cleanUsername}@{domain}");
        }
        
        return validEmails.Count;
    }

    static void Main() {
        Console.Write("Введите адреса почт через пробел: ");
        var emails = new List<string>(Console.ReadLine().Split(' '));
        Console.WriteLine($"Количество уникальных адресов: {CountValidEmails(emails)}");
    }
}

File: lab2_task2/main.rs
use std::collections::HashSet;
use std::io;

fn is_valid_email(email: &str) -> Option<(String, String)> {
    let parts: Vec<&str> = email.split('@').collect();
    if parts.len() != 2 {
        println!("Неверный формат email: отсутствует @ в {}", email);
        return None;
    }
    
    let (username, domain) = (parts[0], parts[1]);
    
    // Проверки имени пользователя
    if username.len() < 6 || username.len() > 30 {
        println!("Имя пользователя должно быть от 6 до 30 символов в {}", email);
        return None;
    }
    
    if username.starts_with('.') || username.ends_with('.') || username.contains("..") {
        println!("Недопустимое расположение точек в имени пользователя в {}", email);
        return None;
    }
    
    for c in username.chars() {
        if !(c.is_ascii_alphanumeric() || c == '.' || c == '*') {
            println!("Недопустимый символ '{}' в имени пользователя в {}", c, email);
            return None;
        }
    }
    
    // Проверки домена
    if domain.is_empty() || domain.starts_with('.') || domain.ends_with('.') || domain.contains("..") {
        println!("Недопустимое расположение точек в домене в {}", email);
        return None;
    }
    
    for c in domain.chars() {
        if !(c.is_ascii_alphanumeric() || c == '.' || c == '-') {
            println!("Недопустимый символ '{}' в домене в {}", c, email);
            return None;
        }
    }
    
    // Обработка звездочки в имени пользователя
    let clean_username = username.replace('.', "");
    let final_username = if clean_username.contains('*') {
        let parts: Vec<&str> = clean_username.split('*').collect();
        if parts[0].len() < 6 {
            println!("Имя пользователя слишком короткое после обработки в {}", email);
            return None;
        }
        parts[0].to_string()
    } else {
        clean_username
    };
    
    Some((final_username, domain.to_string()))
}

fn count_valid_emails(emails: &[&str]) -> usize {
    let mut valid_emails = HashSet::new();
    
    for email in emails {
        if let Some((user, domain)) = is_valid_email(email) {
            valid_emails.insert(format!("{}@{}", user, domain));
        }
    }
    
    valid_emails.len()
}

fn main() {
    println!("Введите адреса почт через пробел:");
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Ошибка чтения ввода");
    
    let emails: Vec<&str> = input.trim().split_whitespace().collect();
    let count = count_valid_emails(&emails);
    
    println!("Количество уникальных адресов: {}", count);
}